### DISCUSSION TITLE
how unit test templating of my dags ?

### DISCUSSION THREAD

**Discussion by raphaelauv** [apache/airflow#26106] [Q&A]
I can already unit test that airflow is capable to parse my dags

but I would like to unit test that my task can be templated by airflow

```python
import logging
import os

from airflow.models import DagBag
from os import path

from airflow.utils.context import Context

src_folder = ""

# this work
def test_dag_loaded():
    dag_parsed = DagBag(dag_folder=path.join(src_folder, 'dags'), include_examples=False)

    for error in dag_parsed.import_errors:
        logging.error(error)
        assert False

# this doesn't work
def test_dag_templating():
    dag_parsed = DagBag(dag_folder=path.join(src_folder, 'dags'), include_examples=False)
    for k, v in dag_parsed.dags.items():
        for a in v.tasks:
            a.render_template_fields(context=Context({
                "toto": "aaaa"
            }))
```

I don't know how to give an airflow context so airflow variable works

because I get this error

```log


self = Undefined, args = (), kwargs = {}

    @internalcode
    def _fail_with_undefined_error(
        self, *args: t.Any, **kwargs: t.Any
    ) -> "te.NoReturn":
        """Raise an :exc:`UndefinedError` when operations are performed
        on the undefined value.
        """
>       raise self._undefined_exception(self._undefined_message)
E       jinja2.exceptions.UndefinedError: 'var' is undefined

```
Thanks

**Comment by potiuk** [comment]
Context is just a dictionary. You can take whatever values you expect to have and simply add them to the dictionary. And in recent Airflow (I think as 2.3, it's even pretty well typed via https://github.com/apache/airflow/blob/main/airflow/utils/context.py#L157 class (which you can instantiate as context) and corresponding context stub: https://github.com/apache/airflow/blob/main/airflow/utils/context.pyi  (so Context class can be even validated with MyPy and autocomplete in IDE's work). 

I think you should just look at the code. Apparently "var" is VariableAccessor which contains either values of json and you can likely look at the airlfow code how to construct it. I think when you go as far as uni-testing individual methods of Airlfow, it only makes sense that you need to look and dive into airflow code to get some details. the template rendering is not "designed" to be black-box.  You can use it, but you need to treat it as white box and simply look at the code (And accept the fact that it might change in the future in newer versions - there is no "guarantees" here nor "formal API - context might change at will so when you decide to unit tests templating you also implicitly decide that you will look at the code and adapt when it changes.

**Reply by potiuk** [reply]
So the only advice I can give - look at the code, and prepare that in the future versions of Airlfow you will have to look again in order to make it works in those future versions.
